### ISSUE TITLE
[no-shadowed-variable] Allow generic types parameters of same name

### ISSUE THREAD

**Issue by RomkeVdMeulen** [palantir/tslint#3799]
### Bug Report

- __TSLint version__: 5.9.1
- __TypeScript version__: 2.8.1
- __Running TSLint via__: gulp-tslint

#### TypeScript code being linted

```ts
export type Partial<T> = {
	[P in keyof T]?: T[P];
};

export interface PartialArray<T> extends Array<Partial<T>> {}

export type PartialWithArrays<T> =
	T extends Array<infer U> ? PartialArray<U> :
	T extends ReadonlyArray<infer U> ? PartialArray<U> :
	Partial<T>;
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "align": false,
    "no-shadowed-variable": true,
  }
}
```

#### Actual behavior

```
ERROR: (no-shadowed-variable) [9, 32]: Shadowed name: 'U'
```

#### Expected behavior
No error.

**Comment by JoshuaKGoldberg** [comment]
### 💀 _**It's time!**_ 💀

**TSLint is deprecated** and no longer accepting pull requests other than security fixes. See #4534. ☠️ 
We recommend you instead use **[typescript-eslint](https://typescript-eslint.io)** to lint your TypeScript code with **ESLint**. ✅

👋 It was a pleasure open sourcing with you!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
